Week3/Homework/routes.js

```diff
@@ -0,0 +1,265 @@
+            day: 'numeric',
+            hour: '2-digit',
+            minute: 'numeric',
+          };
+
+          if (rows[i].complete == 'T') {
+            taskState = 'Done';
+          } else {
+            taskState = 'Not Done';
+          }
+          res.write(
+            `<tr>
+          <td>${rows[i].task_name}</td> 
+          <td>${rows[i].list_id}</td>
+          <td>${rows[i].category_id}</td>
+          <td>${rows[i].deadline.toLocaleString('en-GB', options)}</td>
+          <td>${taskState}</td>
+          <td><button value="${rows[i].task_id}" name="delete_todo_task"/>Delete</button></td>
+          <td><button value="${
+  rows[i].task_name
+}" name="update_todo_task"/>Mark As Complete</button></td>
+          </tr>`,
+          );
+        }
+      }
+
+      res.end('</table></form>');
+    });
+  });
+});
+
+// On path /get_lists Get table: task_list
+
+router.get('/get_lists', (req, res) => {
+  const queryString = 'SELECT * FROM task_list';
+  res.render('viewtodos', (err, html) => {
+    connection.query(queryString, (err, rows, fields) => {
+      if (err) {
+        console.log(`Failed to query @ /get_lists: ${err}`);
+      }
+      console.log('Getting data from database @ /get_lists');
+
+      res.write(html);
+      res.write(`<h1>Todo Lists</h1>
+    <form id="delete_list_form" action="/delete_list" method="POST">
+        <table><tr><th>List Name</th><th>Reminder</th><th></th></tr>`);
+      for (i in rows) {
+        if (Object.prototype.hasOwnProperty.call(rows, i)) {
+          res.write(`<tr><td>${rows[i].list_name}</td>
+        <td>${rows[i].reminder_time.toLocaleString()}</td>
+        <td><button value="${rows[i].list_id}" name="delete_todo_task"/>Delete</button></td></tr>`);
+        }
+      }
+      res.end('</table></form>');
+    });
+  });
+});
+
+// On path /get_users Get table: user
+
+router.get('/get_users', (req, res) => {
+  const queryString = 'SELECT * FROM user';
+  res.render('viewtodos', (err, html) => {
+    connection.query(queryString, (err, rows, fields) => {
+      if (err) {
+        console.log(`Failed to query @ /get_users: ${err}`);
+      }
+      console.log('Getting data from database @ /get_users');
+
+      res.write(html);
+      res.write(`<h1>Users</h1>
+        <table><tr><th>Name</th><th>Email</th><th>Password</th>`);
+      for (i in rows) {
+        if (Object.prototype.hasOwnProperty.call(rows, i)) {
+          res.write(`<tr><td>${rows[i].user_name}</td>
+        <td>${rows[i].user_email}</td>
+        <td>${rows[i].user_password}</td></tr>`);
+        }
+      }
+      res.end('</table>');
+    });
+  });
+});
+
+// On path / add a todo task
+
+router.post('/add_todo', (req, res) => {
+  const todo = req.body.add_todo_input;
+  const deadline = req.body.add_todo_deadline;
+  const tag = req.body.add_todo_tag;
+  const list = req.body.add_todo_list;
+  const queryString = `INSERT INTO task (task_name, date_created, deadline, complete, category_id, list_id) 
```

Comment:
You are vulnerable to attack here. If someone wants to drop your database he can by entering a name like :
```
tt','tt','tt');drop database TodoApplication;
```

You have to use prepared queries with '?'

Reply:
Thank you I think I have done it now